pkg/runner/resume.go

```diff
@@ -0,0 +1,270 @@
+				}
+
+				outputFile, err := os.OpenFile(outputPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+				if err != nil {
+					return errors.Join(packager.ErrCouldNotOpenOutputFile, err)
+				}
+				defer outputFile.Close()
+
+				deviceSize, err := io.Copy(outputFile, inputFile)
+				if err != nil {
+					return errors.Join(snapshotter.ErrCouldNotCopyFile, err)
+				}
+
+				// We need to add a padding like the snapshotter if we're writing to a file instead of a block device
+				if addPadding {
+					if paddingLength := utils.GetBlockDevicePadding(deviceSize); paddingLength > 0 {
+						if _, err := outputFile.Write(make([]byte, paddingLength)); err != nil {
+							return errors.Join(snapshotter.ErrCouldNotWritePadding, err)
+						}
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	goroutineManager := manager.NewGoroutineManager(
+		ctx,
+		&errs,
+		manager.GoroutineManagerHooks{
+			OnAfterRecover: func() {
+				if suspendOnPanicWithError {
+					suspendCtx, cancelSuspendCtx := context.WithTimeout(runner.rescueCtx, rescueTimeout)
+					defer cancelSuspendCtx()
+
+					// Connections need to be closed before creating the snapshot
+					if resumedRunner.acceptingAgent != nil {
+						if e := resumedRunner.acceptingAgent.Close(); e != nil {
+							errs = errors.Join(errs, snapshotter.ErrCouldNotCloseAcceptingAgent, e)
+						}
+					}
+					if resumedRunner.agent != nil {
+						resumedRunner.agent.Close()
+					}
+
+					// If a resume failed, flush the snapshot so that we can re-try
+					if err := resumedRunner.createSnapshot(suspendCtx); err != nil {
+						errs = errors.Join(errs, ErrCouldNotCreateRecoverySnapshot, err)
+					}
+				}
+			},
+		},
+	)
+	defer goroutineManager.Wait()
+	defer goroutineManager.StopAllGoroutines()
+	defer goroutineManager.CreateBackgroundPanicCollector()()
+
+	// We intentionally don't call `wg.Add` and `wg.Done` here since we return the process's wait method
+	// We still need to `defer handleGoroutinePanic()()` here however so that we catch any errors during this call
+	goroutineManager.StartBackgroundGoroutine(func(_ context.Context) {
+		if err := runner.server.Wait(); err != nil {
+			panic(errors.Join(ErrCouldNotWaitForFirecracker, err))
+		}
+	})
+
+	var err error
+	resumedRunner.agent, err = ipc.StartAgentServer(
+		filepath.Join(runner.server.VMPath, snapshotter.VSockName),
+		uint32(agentVSockPort),
+	)
+	if err != nil {
+		panic(errors.Join(snapshotter.ErrCouldNotStartAgentServer, err))
+	}
+
+	resumedRunner.Close = func() error {
+		resumedRunner.agent.Close()
+
+		return nil
+	}
+
+	if err := os.Chown(resumedRunner.agent.VSockPath, runner.hypervisorConfiguration.UID, runner.hypervisorConfiguration.GID); err != nil {
+		panic(errors.Join(ErrCouldNotChownVSockPath, err))
+	}
+
+	{
+		resumeSnapshotAndAcceptCtx, cancelResumeSnapshotAndAcceptCtx := context.WithTimeout(goroutineManager.Context(), resumeTimeout)
+		defer cancelResumeSnapshotAndAcceptCtx()
```

Comment:
Just out of curiosity, but why do we open a new scope here? Was the intention to trigger the `defer` function on exit, because I don't think it will.

Reply:
Nope - just for readability! Didn't want anything else to be able to use the context, but it's ok if it `defer`s later since it just cleans up the RPC loop.